File: src/features/settings/api/extension.ts
import { invoke } from '@tauri-apps/api/core';
import type { ExtensionSettingsTab } from '../pages/extension/types';

interface ExtensionSettingsHostResponse {
  mode: string;
  entry_url: string;
}

interface ExtensionSettingsTabResponse {
  tab_id: string;
  extension_id: string;
  name: string;
  version: string;
  description: string | null;
  path: string;
  enabled: boolean;
  config: string | null;
  host: ExtensionSettingsHostResponse | null;
}

export async function listExtensionSettingsTabs(): Promise<ExtensionSettingsTab[]> {
  const result = await invoke<ExtensionSettingsTabResponse[]>('list_extension_settings_tabs');
  return result.map(mapExtensionSettingsTab);
}

export async function getExtensionSettingsTab(tabId: string): Promise<ExtensionSettingsTab> {
  const result = await invoke<ExtensionSettingsTabResponse>('get_extension_settings_tab', {
    tabId,
  });

  return mapExtensionSettingsTab(result);
}

export async function setExtensionEnabled(extensionId: string, enabled: boolean): Promise<void> {
  await invoke('set_extension_enabled', {
    extensionId,
    enabled,
  });
}

function mapExtensionSettingsTab(result: ExtensionSettingsTabResponse): ExtensionSettingsTab {
  return {
    tabId: result.tab_id,
    extensionId: result.extension_id,
    name: result.name,
    version: result.version,
    description: result.description ?? null,
    path: result.path,
    enabled: Boolean(result.enabled),
    config: result.config ?? null,
    host: result.host
      ? {
          mode: 'iframe',
          entryUrl: result.host.entry_url,
        }
      : null,
  };
}

File: src-tauri/src/models/extension.rs
use serde::{Deserialize, Serialize};
use specta::Type;
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow, Type)]
pub struct Extension {
    pub id: String,
    pub name: String,
    pub version: String,
    pub description: Option<String>,
    pub path: String,
    pub enabled: bool,
    pub config: Option<String>,
    #[specta(type = f64)]
    pub created_at: i64,
    #[specta(type = f64)]
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct CreateExtension {
    pub name: String,
    pub version: Option<String>,
    pub description: Option<String>,
    pub path: String,
    pub config: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ExtensionUpdate {
    pub name: Option<String>,
    pub version: Option<String>,
    pub description: Option<String>,
    pub enabled: Option<bool>,
    pub config: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ExtensionSettingsHostContext {
    pub mode: String,
    pub entry_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct ExtensionSettingsTab {
    pub tab_id: String,
    pub extension_id: String,
    pub name: String,
    pub version: String,
    pub description: Option<String>,
    pub path: String,
    pub enabled: bool,
    pub config: Option<String>,
    pub host: Option<ExtensionSettingsHostContext>,
}
